clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,289 @@
+//===--------- BoundsAnalysis.cpp - Bounds Widening Analysis --------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements a dataflow analysis for bounds widening.
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/Sema/BoundsAnalysis.h"
+
+namespace clang {
+// Legend:
+// n ==> intersection.
+// E ==> belongs.
+// u ==> union.
+
+void BoundsAnalysis::WidenBounds() {
+  assert(Cfg && "expected CFG to exist");
+
+  WorkListTy Worklist;
+  BlockMapTy BlockMap;
+
+  // Add each block to Worklist and create a mapping from Block to
+  // ElevatedCFGBlock. Also update the Gen map for each block.
+  for (const auto *B : PostOrderCFGView(Cfg)) {
+    auto EB = new ElevatedCFGBlock(B);
+    Worklist.insert(EB);
+    BlockMap[B] = EB;
+
+    UpdateGenMap(BlockMap[B], BlockMap);
+  }
+
+  // Dataflow analysis for bounds widening.
+  while (!Worklist.empty()) {
+    auto *EB = Worklist.back();
+    Worklist.pop_back();
+
+    UpdateInMap(EB, BlockMap);
+    auto OldOut = UpdateOutMap(EB);
+
+    // Add the successors of the changed blocks to Worklist, if they do not
+    // already exist in the worklist.
+    if (Differ(OldOut, EB->Out)) {
+      for (const CFGBlock *succ : EB->Block->succs())
+        // Worklist is a SetVector. So it will only insert a key if it does not
+        // already exist.
+        Worklist.insert(BlockMap[succ]);
+    }
+  }
+
+  CollectWidenedBounds(BlockMap);
+}
+
+void BoundsAnalysis::UpdateGenMap(ElevatedCFGBlock *EB, BlockMapTy BlockMap) {
```

Comment:
Normally, a gen set in a dataflow analysis is something that is created by that basic  block.  For example, in constant propagation, a  basic block may contain one or more assignments of variables to constants.  That in turn would cause the basic block to generate a map whose domain is variables definitely known to be constant at the end of the basic block and whose range is the constants values.

That doesn't appear to the intent of your gen set here.   It is unusual that that gen set depends on predecessor gen sets.  I'm also not following the counting that you are doing for each variable declaration in a gen set.  The counts don't appear to correspond to offsets; they appear to be counts of the number if times that a variable was tested.  For example,
```
if (*p) {
  if (*p) {
     // EB->Gen(p) would be 2 here
```
I think what you want to compute for a gen set is whether each predecessor block ends in the same `if (*e)` and that the current block is the true branch if that if.    To widen bounds at the entry to a block, you would walk over all the known bounds, and see if the upper bound of one of them is *e.  If so, then that bound could be widened by 1.